File: CousinsInBinaryTree_dfs.py
#time= O(n)
#space= O(h)
#leetcode=yes 

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    x_parent=None
    x_level=0
    y_parent=None
    y_level=0
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        if(root==None):
            return None
        self.helper(root,x,y,None,0)
        return (self.x_level==self.y_level and self.x_parent!=self.y_parent)
    def helper(self,root,x,y,parent,level):
        if(root==None):
            return 
        if(x==root.val):
            self.x_parent=parent
            self.x_level=level
        if(y==root.val):
            self.y_parent=parent
            self.y_level=level
        self.helper(root.left,x,y,root,level+1)
        self.helper(root.right,x,y,root,level+1)
            
            
            
            
    

File: cousinsInBinaryTreeDFS.java
// Time Complexity: O(n) where n is the number of nodes
// Space Complexity: O(Edge)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int xDepth = -1;
    int yDepth = -1;
    TreeNode xParent;
    TreeNode yParent;
    public boolean isCousins(TreeNode root, int x, int y) {
        dfs(root, x, y, 0, null);
        return xDepth == yDepth && xParent != yParent;
    }
    
    private void dfs(TreeNode root, int x, int y, int depth, TreeNode parent){
        if(root == null) return;
        
        if(root.val == x){
            xDepth = depth;
            xParent = parent;
        }
        if(root.val == y){
            yDepth = depth;
            yParent = parent;
        }
        dfs(root.left, x ,y, depth + 1, root);
        dfs(root.right, x ,y, depth + 1, root);
    }
}